New command line parsing

scordio opened the pull request:
Add new command line parsing feature which works as follow:

```
Usage: csb [seed|synth|veracity|workload] [options] <args>...

  -p, --partitions <num>   Number of RDD partitions [default detected level of parallelism: 8].

Command: seed [options]
Generates a property graph and the probability distributions of its properties starting from two log files.
  -a, --alert-log <path>   Path of the Snort connection log [default: alert].
  -b, --bro-log <path>     Path of the Bro connection log [default: conn.log].
  -l, --log <path>         Output path of the resulting augmented log [default: aug.log].
  -o, --out-graph <path>   Output path prefix of the generated seed graph [default: seed].

Command: synth [ba|kro] [options] <args>...
Synthesizes a graph starting from a seed graph and the probability distributions of its properties.
  -s, --seed-graph <path>  Path prefix of the seed graph [default: seed].
  -l, --log <path>         Path of the augmented log [default: aug.log].
  -o, --out-graph <path>   Output path prefix of the generated synthetic graph [default: synth].
  -x, --exclude-prop       Skip the properties generation [default: false].
  -m, --metrics <metric1,metric2,...>
                           Comma separated list of veracity metrics to execute. Available: degree|in-degree|out-degree|pagerank|all [default:].

Command: synth ba iterations [fraction]
Synthesizes the graph using the Barabási–Albert algorithm.
  iterations               Number of algorithm's iterations.
  fraction                 Fraction of the data extracted at each iteration [default: 0.1].

Command: synth kro iterations [matrix]
Synthesizes the graph using the Kronecker algorithm.
  iterations               Number of algorithm's iterations.
  matrix                   Path of the matrix file representing the starting seed [default: seed.mtx].

Command: veracity <metric1,metric2,...> [seed] [synth]
Executes one or more veracity metrics. Available metrics: degree|in-degree|out-degree|pagerank|all
  <metric1,metric2,...>    Comma separated list of veracity metrics to execute [default: all].
  seed                     Path prefix of the seed graph [default: seed].
  synth                    Path prefix of the synth graph [default: synth].

Command: workload [options] [<workload1,workload2,...>]
Executes one or more workloads. Available workloads: count-vertices|count-edges|degree|in-degree|out-degree|pagerank|bfs|neighbors|in-neighbors|out-neighbors|in-edges|out-edges|connected-components|triangle-count|strongly-connected-components|betweenness-centrality|closeness-centrality|sssp|subgraph-isomorphism|all 
  -b, --backend <value>    Workload backend. Supported: spark|neo4j [default: spark].
  -i, --iterations <num>   Number of algorithm's iterations. Required by: strongly-connected-components|betweenness-centrality [default:].
  -s, --source <id>        The source vertex ID. Required by: bfs|closeness-centrality|sssp [default: random].
  -d, --destination <id>   The destination vertex ID. Required by: bfs [default: random].
  -t, --pattern <path>     Path prefix of the pattern graph. Required by: subgraph-isomorphism [default: pattern].
  -g, --graph <path>       Path prefix of the graph [default: graph].
  <workload1,workload2,...>
                           Comma separated list of workloads to execute [default: all].
```

scordio committed:
Disable all existing cli parsing in Benchmark object, define
new Config case class, define new OptionParser class, create
skeleton of the new cli parsing.

scordio committed:
Rename augLog to outLog, remove seedPrefix as it is replaceable by
outGraphPrefix, add default value for outGraphPrefix in the seed
command context, minor changes to usage descriptions.

scordio committed:
Add synth subcommands and fill veracity command with related parameters;
update Benchmark accordingly.

scordio committed:
Fill workload command with related parameters and update Benchmark
accordingly; add debug option and change default logging level from OFF
to ERROR; remove unused neo4j config statements in the SparkSession
instance.

scordio committed:
Add support for random extraction of source and destination vertex IDs
when not specified.

scordio committed:
Rename outLog to augLog

scordio committed:
Fix augLog assignment in the seed scope, add missing log option in the
synth scope.

scordio committed:
Rename augLog default value, update README to reflect the new command
line.

scordio committed:
Rename run_gendist to run_seed

scordio committed:
Add help option

chmod84 reviewed (changes requested):
Two comments:
1) Although initially we arranged the seed graph to be generated with the data from the "augmented" (bro+snort) log file, we are currently never using the snort labels. I wonder therefore if it is the case to remove it completely so to simplify everything
2) There are no parameters for the connection to Neo4J. Are they hard-coded?

scordio committed:
Add Neo4j connection options in the workload scope, remove static
connection values inside Neo4jWorkload.

scordio committed:
Add asText option to seed and synth commands, add graphLoader and
graphSaver in Config case class, add ComponentFactory class, change
SparkPersistence and Neo4jPersistence from class to object.

scordio committed:
Add loader and saver options, rename exclude-prop to exclude-properties

scordio commented:
> Although initially we arranged the seed graph to be generated with the data from the "augmented" (bro+snort) log file, we are currently never using the snort labels. I wonder therefore if it is the case to remove it completely so to simplify everything

Opened issue #15 to track this change.

> There are no parameters for the connection to Neo4J. Are they hard-coded?

The Neo4j options are now implemented, persistence options have also been added. Updated usage:

````
Usage: csb [seed|synth|veracity|workload] [options] <args>...

  -p, --partitions <num>   Number of RDD partitions [default detected level of parallelism: 32].

Command: seed [options]
Generates a property graph and the probability distributions of its properties starting from two log files.
  -a, --alert-log <path>   Path of the Snort connection log [default: alert].
  -b, --bro-log <path>     Path of the Bro connection log [default: conn.log].
  -l, --log <path>         Output path of the resulting augmented log [default: aug.log].
  -w, --saver <value>      Graph saver. Supported: spark|neo4j [default: spark].
  -o, --out-graph <path>   Output path prefix of the generated seed graph [default: seed].
  -t, --as-text <format>   Save the generated seed graph in a text format. Supported: spark|neo4j [default: none].

Command: synth [ba|kro] [options] <args>...
Synthesizes a graph starting from a seed graph and the probability distributions of its properties.
  -r, --loader <value>     Graph loader. Supported: spark|neo4j [default: spark].
  -s, --seed-graph <path>  Path prefix of the seed graph [default: seed].
  -l, --log <path>         Path of the augmented log [default: aug.log].
  -w, --saver <value>      Graph saver. Supported: spark|neo4j [default: spark].
  -o, --out-graph <path>   Output path prefix of the generated synthetic graph [default: synth].
  -t, --as-text <format>   Save the generated synthetic graph in a text format. Supported: spark|neo4j [default: none].
  -x, --exclude-properties
                           Skip the properties generation [default: false].
  -m, --metrics <metric1,metric2,...>
                           Comma separated list of veracity metrics to execute. Available: degree|in-degree|out-degree|pagerank|all [default:].

Command: synth ba iterations [fraction]
Synthesizes the graph using the Barabási–Albert algorithm.
  iterations               Number of algorithm's iterations.
  fraction                 Fraction of the data extracted at each iteration [default: 0.1].

Command: synth kro iterations [matrix]
Synthesizes the graph using the Kronecker algorithm.
  iterations               Number of algorithm's iterations.
  matrix                   Path of the matrix file representing the starting seed [default: seed.mtx].

Command: veracity [options] <metric1,metric2,...> [seed] [synth]
Executes one or more veracity metrics. Available metrics: degree|in-degree|out-degree|pagerank|all
  -r, --loader <value>     Graph loader. Supported: spark|neo4j [default: spark].
  <metric1,metric2,...>    Comma separated list of veracity metrics to execute [default: all].
  seed                     Path prefix of the seed graph [default: seed].
  synth                    Path prefix of the synth graph [default: synth].

Command: workload [options] [<workload1,workload2,...>]
Executes one or more workloads. Available workloads: count-vertices|count-edges|degree|in-degree|out-degree|pagerank|bfs|neighbors|in-neighbors|out-neighbors|in-edges|out-edges|connected-components|triangle-count|strongly-connected-components|betweenness-centrality|closeness-centrality|sssp|subgraph-isomorphism|all 
  -b, --backend <value>    Workload backend. Supported: spark|neo4j [default: spark].
  --neo4j-url <value>      Neo4j connection URL [default: bolt://localhost].
  --neo4j-username <value>
                           Neo4j username [default: neo4j].
  --neo4j-password <value>
                           Neo4j password [default: neo4j].
  -i, --iterations <num>   Number of algorithm's iterations. Required by: strongly-connected-components|betweenness-centrality [default:].
  -s, --source <id>        The source vertex ID. Required by: bfs|closeness-centrality|sssp [default: random].
  -d, --destination <id>   The destination vertex ID. Required by: bfs [default: random].
  -r, --loader <value>     Graph loader. Supported: spark|neo4j [default: spark].
  -t, --pattern <path>     Path prefix of the pattern graph. Required by: subgraph-isomorphism [default: pattern].
  -g, --graph <path>       Path prefix of the graph [default: graph].
  <workload1,workload2,...>
                           Comma separated list of workloads to execute [default: all].
````

scordio committed:
Add comments to ComponentFactory

scordio committed:
Rename workload and getWorkload to workloadEngine and getWorkloadEngine,
move metric parsing in ComponentFactory.

scordio committed:
Add name field to the Veracity trait.
